test/idleness_test.go

```diff
@@ -0,0 +1,481 @@
+		if l := len(tcs); l != 1 {
+			return fmt.Errorf("when looking for channelz trace event with description %q, found %d top-level channels, want 1", wantDesc, l)
+		}
+		if tcs[0].Trace == nil {
+			return fmt.Errorf("when looking for channelz trace event with description %q, no trace events found for top-level channel", wantDesc)
+		}
+
+		for _, e := range tcs[0].Trace.Events {
+			if strings.Contains(e.Desc, wantDesc) {
+				return nil
+			}
+		}
+	}
+	return fmt.Errorf("when looking for channelz trace event with description %q, %w", wantDesc, ctx.Err())
+}
+
+// channelzTraceEventNotFound looks up the top-channels in channelz (expects a
+// single one), and verifies that there is no trace event on the channel
+// matching the provided description string.
+func channelzTraceEventNotFound(ctx context.Context, wantDesc string) error {
+	sCtx, sCancel := context.WithTimeout(ctx, defaultTestShortTimeout)
+	defer sCancel()
+
+	err := channelzTraceEventFound(sCtx, wantDesc)
+	if err == nil {
+		return fmt.Errorf("found channelz trace event with description %q, when expected not to", wantDesc)
+	}
+	if !errors.Is(err, context.DeadlineExceeded) {
+		return err
+	}
+	return nil
+}
+
+// Tests the case where channel idleness is disabled by passing an idle_timeout
+// of 0. Verifies that a READY channel with no RPCs does not move to IDLE.
+func (s) TestChannelIdleness_Disabled_NoActivity(t *testing.T) {
+	// Setup channelz for testing.
+	czCleanup := channelz.NewChannelzStorageForTesting()
+	t.Cleanup(func() { czCleanupWrapper(czCleanup, t) })
+
+	// Create a ClientConn with idle_timeout set to 0.
+	r := manual.NewBuilderWithScheme("whatever")
+	dopts := []grpc.DialOption{
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+		grpc.WithResolvers(r),
+		grpc.WithIdleTimeout(0), // Disable idleness.
+		grpc.WithDefaultServiceConfig(`{"loadBalancingConfig": [{"round_robin":{}}]}`),
+	}
+	cc, err := grpc.Dial(r.Scheme()+":///test.server", dopts...)
+	if err != nil {
+		t.Fatalf("grpc.Dial() failed: %v", err)
+	}
+	t.Cleanup(func() { cc.Close() })
+
+	// Start a test backend and push an address update via the resolver.
+	backend := stubserver.StartTestService(t, nil)
+	t.Cleanup(func() { backend.Stop() })
+	r.UpdateState(resolver.State{Addresses: []resolver.Address{{Addr: backend.Address}}})
+
+	// Veirfy that the ClientConn moves to READY.
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	for state := cc.GetState(); state != connectivity.Ready; state = cc.GetState() {
+		if !cc.WaitForStateChange(ctx, state) {
+			t.Fatal("Timeout when waiting for channel to switch to READY")
+		}
+	}
+
+	// Veirfy that the ClientConn stay in READY.
+	sCtx, sCancel := context.WithTimeout(ctx, 3*defaultTestShortIdleTimeout)
+	defer sCancel()
+	if cc.WaitForStateChange(sCtx, connectivity.Ready) {
+		t.Fatalf("Connectivity state changed to %q when expected to stay in READY", cc.GetState())
+	}
```

Comment:
Use these instead!

```go
func awaitState(ctx context.Context, t *testing.T, cc *grpc.ClientConn, stateWant connectivity.State)
func awaitNotState(ctx context.Context, t *testing.T, cc *grpc.ClientConn, stateDoNotWant connectivity.State)
```

Reply:
Or factor out similarly, next to these (in `clientconn_state_transition_test.go`), since this particular test doesn't seem to be able to reuse those.

Reply:
Done. And also, added a `awaitNoStateChange` alongside the existing ones. At some point, we can move them to the `testutils` package.